server/chroma/src/main.rs

```diff
@@ -1,100 +1,51 @@
 extern crate core;
 
-use crate::config::Config;
-use crate::routes::appdata::{AlbumIdCache, AppData, Ratelimits, SessionIdCache, WebData};
-use crate::routes::routable::Routable;
-use actix_cors::Cors;
-use actix_web::{web, App, HttpServer};
-use cabbage::KoalaApi;
-use color_eyre::eyre::Error;
-use color_eyre::Result;
-use dal::database::Database;
-use dal::storage_engine::{S3Config, Storage};
+use std::process::ExitCode;
+
 use dotenv::dotenv;
-use noiseless_tracing_actix_web::NoiselessRootSpanBuilder;
-use std::time::Duration;
-use tracing::{info, warn};
-use tracing_actix_web::TracingLogger;
+use tracing::level_filters::LevelFilter;
+use tracing::{error, warn};
 use tracing_subscriber::fmt::layer;
 use tracing_subscriber::layer::SubscriberExt;
 use tracing_subscriber::util::SubscriberInitExt;
 use tracing_subscriber::EnvFilter;
 
+mod app;
 mod config;
 mod routes;
 
 #[tokio::main]
-async fn main() -> Result<()> {
-    dotenv().ok();
-
-    install_tracing();
+async fn main() -> ExitCode {
+    // Try to load the environment variables from the .env file
+    let dotenv_err = dotenv().err();
 
-    info!("Starting");
+    // Initialize the tracing logger
+    init_tracing();
 
-    info!("Parsing config");
-    let config = Config::parse()?;
-
-    if !config.validate() {
-        return Err(Error::msg("Config is not valid."));
+    // Check if the dot env was loaded correctly, otherwise send a warning
+    if let Some(err) = dotenv_err {
+        warn!("failed to load .env file: {:#}", err);
     }
 
-    if !config.service_tokens.is_empty() {
-        warn!("There are service tokens configured, Make sure these are, and stay, confidential!");
+    // Run the server
+    if let Err(err) = app::run().await {
+        error!("{}", err);
+        err.chain()
+            .skip(1)
+            .for_each(|cause| error!("because: {}", cause));
+        ExitCode::from(1)
```

Comment:
Why not return an `Ok` or `Err` case here? Let the runtime figure out the appropriate error code to return?

Reply:
As returning `Ok` or `Err` will print the message twice; once using tracing (done manually) and then a second time by `eprintln!()` as rust automatically prints the error to `stderr` when returning an `Err` from `main`. Furthermore, rust (and anyhow) completely ignore exit codes and only return `libc::EXIT_SUCCESS` and `libc::EXIT_FAILURE`, respectively `0` and `1`.[^1]

[^1]: https://doc.rust-lang.org/std/process/trait.Termination.html